sql/spanner/migrate.go

```diff
@@ -0,0 +1,315 @@
+	}
+	return nil
+}
+
+func (s *state) indexParts(b *sqlx.Builder, parts []*schema.IndexPart) {
+	b.Wrap(func(b *sqlx.Builder) {
+		b.MapComma(parts, func(i int, b *sqlx.Builder) {
+			switch part := parts[i]; {
+			case part.C != nil:
+				b.Ident(part.C.Name)
+			case part.X != nil:
+				b.WriteString(sqlx.MayWrap(part.X.(*schema.RawExpr).X))
+			}
+			if parts[i].Desc {
+				b.P("DESC")
+			}
+		})
+	})
+}
+
+func (s *state) dropForeignKeys(t *schema.Table, foreignKeys ...*schema.ForeignKey) error {
+	rs := &state{conn: s.conn}
+	if err := rs.addForeignKeys(t, foreignKeys...); err != nil {
+		return err
+	}
+	for i := range rs.Changes {
+		s.append(&migrate.Change{
+			Cmd:     rs.Changes[i].Reverse,
+			Reverse: rs.Changes[i].Cmd,
+			Comment: fmt.Sprintf("drop foreignKey %q from table: %q", foreignKeys[i].Symbol, t.Name),
+		})
+	}
+	return nil
+}
+
+func (s *state) addForeignKeys(t *schema.Table, foreignKeys ...*schema.ForeignKey) error {
+	for _, fk := range foreignKeys {
+		b := Build("ALTER TABLE")
+		b.Ident(t.Name)
+		b.P("ADD")
+		s.fks(b, foreignKeys...)
+		fkName := fk.Symbol
+		// TODO: derive fk name if we don't have one in Symbol.
+		s.append(&migrate.Change{
+			Cmd:    b.String(),
+			Source: &schema.AddForeignKey{F: fk},
+			Reverse: Build("ALTER TABLE").
+				Ident(t.Name).
+				P(fmt.Sprintf("DROP CONSTRAINT %v", fkName)).
+				String(),
+			Comment: fmt.Sprintf("create foreignKey %q to table: %q", fk.Symbol, t.Name),
+		})
+	}
+	return nil
+}
+
+func (s *state) fks(b *sqlx.Builder, fks ...*schema.ForeignKey) {
+	b.MapComma(fks, func(i int, b *sqlx.Builder) {
+		fk := fks[i]
+		if fk.Symbol != "" {
+			b.P("CONSTRAINT").Ident(fk.Symbol)
+		}
+		b.P("FOREIGN KEY")
+		b.Wrap(func(b *sqlx.Builder) {
+			b.MapComma(fk.Columns, func(i int, b *sqlx.Builder) {
+				b.Ident(fk.Columns[i].Name)
+			})
+		})
+		if fk.RefTable != nil {
+			b.P("REFERENCES").Ident(fk.RefTable.Name)
+			b.Wrap(func(b *sqlx.Builder) {
+				b.MapComma(fk.RefColumns, func(i int, b *sqlx.Builder) {
+					b.Ident(fk.RefColumns[i].Name)
+				})
+			})
+		}
+	})
+}
+
+func (s *state) append(c *migrate.Change) {
+	s.Changes = append(s.Changes, c)
+}
+
+// checks writes the CHECK constraint to the builder.
+func check(b *sqlx.Builder, c *schema.Check) {
+	expr := c.Expr
+	// Expressions should be wrapped with parens.
+	if t := strings.TrimSpace(expr); !strings.HasPrefix(t, "(") || !strings.HasSuffix(t, ")") {
+		expr = "(" + t + ")"
+	}
+	if c.Name != "" {
+		b.P("CONSTRAINT").Ident(c.Name)
+	}
+	b.P("CHECK", expr)
+}
+
+// Build instantiates a new builder and writes the given phrase to it.
+func Build(phrase string) *sqlx.Builder {
+	b := &sqlx.Builder{QuoteChar: '`'}
+	return b.P(phrase)
+}
+
+func defaultValue(c *schema.Column) (string, error) {
+	switch x := c.Default.(type) {
+	case *schema.Literal:
+		return sqlx.SingleQuote(x.V)
+	case *schema.RawExpr:
+		return x.X, nil
+	default:
+		return "", fmt.Errorf("unexpected default value type: %T", x)
+	}
+}
```

Comment:
Unrelated to this change, but it feels like some migration parts can move to `sqlx`.